File: examples/java/product/src/main/java/dev/angzarr/examples/product/CommandValidationException.java
package dev.angzarr.examples.product;

import io.grpc.Status;

public class CommandValidationException extends Exception {
    private final Status.Code statusCode;

    public CommandValidationException(Status.Code statusCode, String message) {
        super(message);
        this.statusCode = statusCode;
    }

    public static CommandValidationException invalidArgument(String message) {
        return new CommandValidationException(Status.Code.INVALID_ARGUMENT, message);
    }

    public static CommandValidationException failedPrecondition(String message) {
        return new CommandValidationException(Status.Code.FAILED_PRECONDITION, message);
    }

    public Status.Code getStatusCode() {
        return statusCode;
    }
}

File: examples/kotlin/fulfillment/src/main/kotlin/dev/angzarr/examples/fulfillment/CommandValidationException.kt
package dev.angzarr.examples.fulfillment

import io.grpc.Status

class CommandValidationException(
    message: String,
    val statusCode: Status.Code
) : Exception(message) {

    companion object {
        fun invalidArgument(message: String): CommandValidationException =
            CommandValidationException(message, Status.Code.INVALID_ARGUMENT)

        fun failedPrecondition(message: String): CommandValidationException =
            CommandValidationException(message, Status.Code.FAILED_PRECONDITION)
    }
}
